platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java

```diff
@@ -49,7 +70,117 @@ public static String formatNodeName(final long nodeId) {
      * @return its Hash
      */
     @NonNull
-    public static Hash hash(@NonNull final Roster roster) {
+    public static Hash hashOf(@NonNull final Roster roster) {
         return PBJ_RECORD_HASHER.hash(roster, Roster.PROTOBUF);
     }
+
+    /**
+     * Constructs the initial roster based on the given software version and initial state.
+     * The initial roster is constructed by adopting the candidate roster if a software upgrade is detected.
+     * Otherwise, the initial roster is created from the address book.
+     *
+     * @param version      the software version
+     * @param initialState the initial state
+     * @param addressBook  the address book
+     * @return the initial roster
+     */
+    public static @NonNull Roster constructInitialRoster(
+            @NonNull final SoftwareVersion version,
+            @NonNull final ReservedSignedState initialState,
+            @NonNull final AddressBook addressBook) {
+        final boolean softwareUpgrade = detectSoftwareUpgrade(version, initialState.get());
+        final MerkleRoot merkleRoot = initialState.get().getState();
+        final WritableRosterStore rosterStore = merkleRoot.getWritableRosterStore();
+        final Roster candidateRoster = rosterStore.getCandidateRoster();
+
+        return constructRoster(
+                rosterStore,
+                softwareUpgrade,
+                candidateRoster,
+                addressBook,
+                initialState.get().getRound());
+    }
+
+    /**
+     * Creates a new roster from the bootstrap address book.
+     * @param addressBook the address book
+     * @return a new roster
+     */
+    @NonNull
+    public static Roster createRoster(@NonNull final AddressBook addressBook) {
+        Objects.requireNonNull(addressBook, "The AddressBook must not be null.");
+        final List<RosterEntry> rosterEntries = new ArrayList<>(addressBook.getSize());
+        for (int i = 0; i < addressBook.getSize(); i++) {
+            final NodeId nodeId = addressBook.getNodeId(i);
+            final Address address = addressBook.getAddress(nodeId);
+
+            final RosterEntry rosterEntry = RosterUtils.toRosterEntry(address, nodeId);
+            rosterEntries.add(rosterEntry);
+        }
+        return Roster.newBuilder().rosterEntries(rosterEntries).build();
+    }
+
+    /**
+     * Converts an address to a roster entry.
+     *
+     * @param address the address to convert
+     * @param nodeId  the node ID to use for the roster entry
+     * @return the roster entry
+     */
+    static RosterEntry toRosterEntry(@NonNull final Address address, @NonNull final NodeId nodeId) {
```

Comment:
Does this need to be package-private?

Reply:
On a second look, no. I've reworked the tests to no longer need it. Thanks!